### ISSUE TITLE
Add support for window rules

### ISSUE THREAD

**Issue by txtyash** [kmonad/kmonad#393]
Can we have support for window rules?!!
I beg you! Please add support for window rules. It'd be the best thing ever. My workflow is already amazing with kmonad but if we could have window rules then it'd be amazing. 

Specific layers for specific applications/windows. It's not a niche situation. If you add this feature then it'll change everything.

**Comment by slotThe** [comment]
Not trying to destroy your hopes here, but I'm fairly certain that this it out-of-scope for kmonad.  David has stated multiple times that he wants it to stay as close to the raw keycodes as possible, and window-aware keybindings are extremetly environment dependent (active keyboard grabs for X11, god knows what for Wayland/MacOS/Windows).  Ideally, this is a feature that xmshould be supported by your window manager (or the window manager of your DE)

**Comment by zim0369** [comment]
That is disappointing. Why don't you try to add support for just x11 and see how good/bad it is?

**Comment by slotThe** [comment]
Thinking about this a bit more, I'm not so sure anymore this is even possible as a third party application[^1].  We can't _really_ issue an active grab because then key events would not be passed through to anything else (even the wm) and we can't issue a passive grab either because these only fire when no parent window has a passive grab on that same key combination (and window managers usually issues a passive grab to the root window).  It is really something that should be handled by the wm.

I can't imagine this being easier for proprietary operating systems or Wayland.  Even if it worked, having our own X11 event loop just for this[^2] just seems wildly out-of-scope for a program that, _a priori_, just reads a few bytes from uinput (or equivalent sources).

[^1]: This is all X11 talk now.
[^2]: I shudder to think how it would work on Wayland; would we need to implement a compositor just for this?

**Comment by MuhammedZakir** [comment]
I don't think it's _that_ complicated. What we want is to detect the frontmost application window, or in other words, currently focused window. Then, change the key remapping according to the title or class of that window.

```clojure
(defalias
  caps (condi (app :title "*nvim*" :class "*wezterm*")
     (


`Karabiner-Elements` already have this. There are other apps for Linux X11 with the ability to detect focused window. For example, `xdotool` has `getactivewindow` option.. So I think it won't be hard to implement this for Linux X11 and macOS. I don't know about Windows but it should provide some API to get the currently focused window. But, I don't have hopes for Wayland [1].

[1] In "wayland support" issue in repo of `espanso`, they mentioned there is no option to detect the frontmost application window. This maybe possible in the future. Otherwise, many good apps will be crippled in Wayland.

**Comment by slotThe** [comment]
> I don't think it's _that_ complicated. What we want is to detect the frontmost application window, or in other words, currently focused window. 

This is relatively trivial to do (pretty much all wm's support EWMH in some capacity at this point, so we'd just have to read `_NET_ACTIVE_WINDOW`.  My point was that what do we do if the wm has already grabbed the respective keys?  We would have to have at least a small event loop to get that information and report it to the user.  
 
> ```clojure
> (defalias
>   caps (condi (app :title "*nvim*" :class "*wezterm*")
>               (tap-next-release esc lctl)
>               (caps))
> )

However, your point here is well taken; what we are _really_ interested in is not necessarily keys that a wm might grab (say, the combination `M-<RET>`) but more single keys that no actual keyboard grab should be active on (say, the key `c`).  In that sense we would side-step the whole "passive grabs bubble up" problem and could just issue these (at least for X11).

**Comment by MuhammedZakir** [comment]
> [...] My point was that what do we do if the wm has already grabbed the respective keys? [...]

Why should we care about keys bound by wm? I thought KMonad intercepts key events before wm does. Am I wrong?

**Comment by zim0369** [comment]
> > [...] My point was that what do we do if the wm has already grabbed the respective keys? [...]
> 
> Why should we care about keys bound by wm? I thought KMonad intercepts key events before wm does. Am I wrong?

IG what he meant was that if you've already using a binding for your wm then you won't be able to use it elsewhere for a different action, even with window rules.

**Comment by MuhammedZakir** [comment]
> IG what he meant was that if you've already using a binding for your wm then you won't be able to use it elsewhere for a different action, even with window rules.

That's correct. But... I'm sorry, I still don't get what the problem is. :-( Window-related rule in a key remapper is to decide what to do for an input event when a particular window is in focus. How that is interpreted by the system is beyond key remapper.

Maybe I am missing something? It would be nice if someone could provide an example/scenario where we need to care about keys bound by wm.

**Comment by slotThe** [comment]
> > [...] My point was that what do we do if the wm has already grabbed the respective keys? [...]
> 
> Why should we care about keys bound by wm? I thought KMonad intercepts key events before wm does. Am I wrong?

Nono, you are totally right; I was the confused one because I forgot how great kmonad is :)  Sorry for the noise

**Comment by alikthename** [comment]
the ability to send some sort of inter-app trigger to change a layout could make a significant user experience difference.

**Comment by david-janssen** [comment]
> the ability to send some sort of inter-app trigger to change a layout could make a significant user experience difference.

Agreed :-). It's on my pile of things, about a third of the way from the top.

**Comment by berenddeboer** [comment]
As I was investigating kmonad to replace my set of Karabiner/SpaceLauncher/iterm2 profiles, I think this feature is essential on a Mac.

My use case: I want to use Ctrl+C for copy, Ctrl+V for paste. OK, just remap these keys to Cmd+C and Cmd+V on map.

However inside iterm2 I run tmux + emacs. There Ctrl+C should remain Ctrl+C.

If I go into my AWS Linux AWS Workspace, same: Ctrl+C needs to stay Ctrl+C. I can't see how I could configure this without having a facility to change the layout per focused window. 

But please let me know if I'm missing something.

**Comment by david-janssen** [comment]
> But please let me know if I'm missing something.

No, I think you're spot on. The keyboard needs to do different things dependent on the state of the operating system (either windows or focus or whatever). This is simply not doable at the moment with KMonad, although I'm hoping to add support for this.

**Comment by yyy33** [comment]
I know that working with awesome wm can make your wishes come true

**Comment by cheyngoodman** [comment]
Without this feature, I need to consider [keyd](https://github.com/rvaiya/keyd) or [xremap](https://github.com/k0kubun/xremap) as an alternative.

**Comment by derelbenkoenig** [comment]
What about a way to trigger kmonad presses via IPC? Like set a "button" that is mapped not to any key on the keyboard but could be triggered with some kind of `kmonad-cmd foo` and have a separate program listen for window focus events in the WM, and send the appropriate command to kmonad to change layers?

**Comment by jokesper** [comment]
I like the Idea but hadn't gotten around to implementing it. The problem I see, is cross compatibility. Currently you can use your same config on linux as on mac / windows. The only problem is the `input` and `output` block, you have to replace.
Events as input would add more complexity to the operating system specific stuff.

I think something like:
- #819 
- #847
- #562

may be better ideas to implement

**Comment by derelbenkoenig** [comment]
Oh, yeah. I noticed #136 because it was one of the issues findable when I was searching through issues for things that could be related to window focus events. It says "signals" specifically but I was thinking more along the lines of some sort of IPC in general

**Comment by jokesper** [comment]
The thing preventing me is that I can't find a good solution for a socket.
I would like an option to specify a socket path, which should be an m-to-n connection.
Different instances of KMonad probably specify the same socket. (e. g.: I run the same config for all keyboards with an `-i` to overwrite which one).
I would also imagine multiple programs listening to layer-change events.

I could think of two solutions, which both have a lot of drawbacks.
1. Connect to a DBus session, this has the problem of restricting which user kmonad should run as. I.e. it's trivial (and not recommended) to run KMonad as `root`. To which dbus session should it connect. Furthermore the nixos default, is sandboxing `kmonad` in it's own user. This is very nice, since it isn't running as root, but other applications may not arbitrarly grab keyboards.
2. Having a `kmonad-daemon` to which instances connect. In the simple case it's effectively dbus. in the monolithic case where we have one Daemon which does everything and kmonad requests to add new inputs, it would be quite a rewrite.

So what we need, is a solution for cross user, n-to-m connections over a single id. Preferrably this wouldn't depend on specific users but groups.

If you have any Ideas, please let me now
